snapcraft/internal/lxd.py

```diff
@@ -395,9 +395,10 @@ def _finish(self):
             process.terminate()
 
     def execute(self, step='snap', args=None):
-        # clean with no parts deletes the container
-        if step == 'clean' and args == ['--step', 'pull']:
-            if self._get_container_status():
+        if step == 'clean':
+            # we don't execute clean here but rely on CLI to do it
+            # clean with no parts deletes the container
+            if self._get_container_status() and args == ['--step', 'pull']:
```

Comment:
I think re-purposing this function to clean is both confusing (`clean` is not a lifecycle step, and then one sees `--step pull` and gets _more_ confused) and is causing the diff of this PR to be larger than it needs to be (changing the step from `None` to `pull` everywhere). How difficult would it be to simply add a `clean` function here that accepts a step and args?

Reply:
Eh, I changed my mind on defaulting step to pull at a higher level, otherwise you'll need to treat `None` and `'pull'` as equivalent here. However, I would still like to see a `clean` function here instead of `execute`.

Reply:
I reverted the default change now. It's not essential to the PR, it just seemed like a simplification.

And I replaced the subclass' execute logic with a delete method - now lifecycle remains in charge and reading lxd.py doesn't require any knowledge of what lifecycle does with it.